# 1. Imports

In [1]:
import pandas as pd
import numpy as np
from warnings import filterwarnings
filterwarnings('ignore')

# 2. Data

In [2]:
apps=pd.read_csv('googleplaystore.csv')
reviews=pd.read_csv('googleplaystore_user_reviews.csv')

In [3]:
apps.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [4]:
reviews.head()

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000

# 3. Basic information about apps

In [5]:
apps.shape

(10841, 13)

In [6]:
apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [7]:
apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [8]:
apps.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

### Inferences so far..
1. There are 10,841 rows and 13 columns. We should check for duplicate data as well.
2. There are some missing values in the rating column. This might indicate that these apps are relatively new.
3. Data types of many columns are not appropriate. We might need to change that in future.

In [9]:
apps.duplicated().sum()

483

# 4. Data cleaning

### 4.1 Let's drop duplicate rows

In [10]:
apps.drop_duplicates(inplace=True)

In [11]:
apps.shape

(10358, 13)

In [12]:
apps.nunique()

App               9660
Category            34
Rating              40
Reviews           6002
Size               462
Installs            22
Type                 3
Price               93
Content Rating       6
Genres             120
Last Updated      1378
Current Ver       2832
Android Ver         33
dtype: int64

### 4.2 Let's change the datatypes to get a clear description of the data.

In [13]:
apps['Reviews']=apps['Reviews'].str.replace('M','')

In [14]:
print (sum(apps.Price.str.match('^$\d+\.?\d*$'))
       + sum(apps.Price == '0')
      )

9592


In [15]:
apps=apps.astype({'Reviews': float})

In [16]:
apps=apps.astype({'Reviews': int })

In [17]:
apps['Price'].nunique()

93

In [18]:
from copy import deepcopy
# copy the desired column to a new variable
#   and work on this new variable
#   to avoid unnecessary errors for the main data frame.
apps_price = deepcopy(apps.Price)
# get the indexes of the rows that are 
#   not ready to be converted to float.
index_price_dollar = (
    apps.Price.str.match('^$\d+\.?\d*$')
)
# do the conversion.
apps_price[index_price_dollar] = (
    apps_price[index_price_dollar].str[1:].astype(float)
)
# apply the change to the main data frame.
apps.Price = apps_price
# delete temporary variables
del apps_price

In [19]:
apps['Price'].unique()

array(['0', '$4.99', '$3.99', '$6.99', '$1.49', '$2.99', '$7.99', '$5.99',
       '$3.49', '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49',
       '$10.00', '$24.99', '$11.99', '$79.99', '$16.99', '$14.99',
       '$1.00', '$29.99', '$12.99', '$2.49', '$10.99', '$1.50', '$19.99',
       '$15.99', '$33.99', '$74.99', '$39.99', '$3.95', '$4.49', '$1.70',
       '$8.99', '$2.00', '$3.88', '$25.99', '$399.99', '$17.99',
       '$400.00', '$3.02', '$1.76', '$4.84', '$4.77', '$1.61', '$2.50',
       '$1.59', '$6.49', '$1.29', '$5.00', '$13.99', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$19.90', '$8.49', '$1.75',
       '$14.00', '$4.85', '$46.99', '$109.99', '$154.99', '$3.08',
       '$2.59', '$4.80', '$1.96', '$19.40', '$3.90', '$4.59', '$15.46',
       '$3.04', '$4.29', '$2.60', '$3.28', '$4.60', '$28.99', '$2.95',
       '$2.90', '$1.97', '$200.00', '$89.99', '$2.56', '$30.99', '$3.61',
       '$394.99', '$1.26', 'Everyone', '$1.20', '$1.04'], dtype=object)

In [20]:
apps[apps['Price']=='Everyone']

App Category  Rating  Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0        3   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

In [21]:
apps.drop(labels=10472, inplace=True)
apps['Price'] = apps['Price'].apply(lambda x: str(x).replace('$', '') if '$' in str(x) else str(x))
apps['Price'] = apps['Price'].apply(lambda x: float(x))

In [22]:
apps['Installs']=apps['Installs'].str.replace('+','')

In [23]:
apps['Installs']=apps['Installs'].str.replace(',','')

In [24]:
apps['Installs']=apps['Installs'].str.replace('Free','0')

In [25]:
apps['Installs']=apps['Installs'].astype(int)

In [26]:
apps['Size']=apps['Size'].str.replace('1,000+','1000')

In [27]:
sum((apps['Size'].str.match('\d+\.?\d*M').sum(),
     apps['Size'].str.match('\d+\.?\d*k').sum(),
     apps['Size'].str.match('Varies with device').sum()))

10357

In [28]:
# copy the desired column to a new variable
#   and work on this new variable
#   to avoid unnecessary errors for the main data frame.
apps_size = deepcopy(apps['Size'])
# get the indexes of rows for each type of size.
index_size_M = apps_size.str.endswith('M')
index_size_k = apps_size.str.endswith('k')
index_size_unknown = apps_size.str.match('Varies with device')
# replace the values of the copy.
apps_size[index_size_M] = (
    apps_size[index_size_M].apply(
        lambda x: float(x[:-1])
    )
)
apps_size[index_size_k] = (
    apps_size[index_size_k].apply(
        lambda x : float(x[:-1]) / 1024
    )
)
apps_size[index_size_unknown] = np.nan
# apply the replaced value to the main data frame.
apps['Size'] = apps_size.astype(float)
# make a new column to emphasize the apps with
#   variable size.
apps['Variable Size'] = (
    apps['Size'].isna().astype(int)
)
# delete temporary variables.
del apps_size
del index_size_M, index_size_k, index_size_unknown

In [29]:
apps.drop(columns='Variable Size', axis=1, inplace=True)

In [30]:
apps.dtypes

App                object
Category           object
Rating            float64
Reviews             int64
Size              float64
Installs            int64
Type               object
Price             float64
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

### 4.3 Now that our datatypes look good, we can move on to fill missing values.

In [31]:
apps.isnull().sum()

App                  0
Category             0
Rating            1465
Reviews              0
Size              1526
Installs             0
Type                 1
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

In [32]:
apps['Size'].fillna(apps['Size'].median(),inplace=True)
apps['Rating'].fillna(apps['Rating'].median(),inplace=True)

### There are some more missing values, but those columns don't appear to be that important for now. 

In [33]:
apps.describe().round(2)

Rating      Reviews      Size      Installs     Price
count  10357.00     10357.00  10357.00  1.035700e+04  10357.00
mean       4.20    405904.61     20.07  1.415776e+07      1.03
std        0.49   2696777.84     21.02  8.023955e+07     16.28
min        1.00         0.00      0.01  0.000000e+00      0.00
25%        4.10        32.00      5.70  1.000000e+03      0.00
50%        4.30      1680.00     13.00  1.000000e+05      0.00
75%        4.50     46416.00     26.00  1.000000e+06      0.00
max        5.00  78158306.00    100.00  1.000000e+09    400.00

# 5. EDA
Let's check the columns and datatypes first, so that we don't miss anything.

In [34]:
apps.dtypes

App                object
Category           object
Rating            float64
Reviews             int64
Size              float64
Installs            int64
Type               object
Price             float64
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

### 5.1 Let's first groupby the data based on categorical columns to check for any pattern.

In [35]:
apps.groupby(by='Category')['App'].count()

Category
ART_AND_DESIGN           65
AUTO_AND_VEHICLES        85
BEAUTY                   53
BOOKS_AND_REFERENCE     230
BUSINESS                427
COMICS                   60
COMMUNICATION           366
DATING                  196
EDUCATION               130
ENTERTAINMENT           111
EVENTS                   64
FAMILY                 1943
FINANCE                 360
FOOD_AND_DRINK          124
GAME                   1121
HEALTH_AND_FITNESS      306
HOUSE_AND_HOME           80
LIBRARIES_AND_DEMO       85
LIFESTYLE               373
MAPS_AND_NAVIGATION     137
MEDICAL                 408
NEWS_AND_MAGAZINES      264
PARENTING                60
PERSONALIZATION         388
PHOTOGRAPHY             322
PRODUCTIVITY            407
SHOPPING                224
SOCIAL                  280
SPORTS                  351
TOOLS                   843
TRAVEL_AND_LOCAL        237
VIDEO_PLAYERS           175
WEATHER                  82
Name: App, dtype: int64

In [36]:
apps.groupby(by='Category').mean().round(2)

Rating     Reviews   Size     Installs  Price
Category                                                          
ART_AND_DESIGN         4.36    26376.00  12.40   1912893.85   0.09
AUTO_AND_VEHICLES      4.21    13690.19  19.21    625061.31   0.16
BEAUTY                 4.28     7476.23  13.71    513151.89   0.00
BOOKS_AND_REFERENCE    4.34    95100.99  13.16   8332476.42   0.52
BUSINESS               4.18    28941.85  13.81   2022634.34   0.41
COMICS                 4.16    56387.93  13.41    934769.17   0.00
COMMUNICATION          4.18  1642823.91  12.05  65989825.82   0.23
DATING                 4.03    28292.84  15.57   1053755.65   0.16
EDUCATION              4.38   178196.15  18.23   4107323.08   0.14
ENTERTAINMENT          4.14   428565.01  18.82  22123063.06   0.07
EVENTS                 4.40     2515.91  13.86    249580.64   1.72
FAMILY                 4.20   204205.85  26.88   5168138.19   1.25
FINANCE                4.15    47221.92  17.17   2139857.59   8.06
FOOD_AND_DRINK         4.18    61867.55  20.16   2079828.64   0.07
GAME                   4.28  1262744.56  40.77  28139183.24   0.26
HEALTH_AND_FITNESS     4.27   100801.26  19.78   4447785.99   0.21
HOUSE_AND_HOME         4.19    34934.65  15.30   1565155.76   0.00
LIBRARIES_AND_DEMO     4.21    12201.39  11.15    741128.35   0.01
LIFESTYLE              4.13    34371.51  14.68   1433843.27   6.33
MAPS_AND_NAVIGATION    4.08   223790.18  15.85   5286729.12   0.20
MEDICAL                4.21     3423.42  18.95    103441.61   2.50
NEWS_AND_MAGAZINES     4.16   144870.73  12.72  20428855.15   0.02
PARENTING              4.30    15972.18  21.56    525351.83   0.16
PERSONALIZATION        4.33   193796.81  11.44   5346636.04   0.39
PHOTOGRAPHY            4.19   634464.01  15.89  30190210.11   0.38
PRODUCTIVITY           4.22   251976.65  12.91  30621846.12   0.62
SHOPPING               4.26   423799.83  15.34  11488164.22   0.02
SOCIAL                 4.26  1905631.53  15.81  44692385.36   0.06
SPORTS                 4.24   186104.58  22.43   4354913.10   0.28
TOOLS                  4.08   324062.92   9.29  13585731.81   0.32
TRAVEL_AND_LOCAL       4.12   234452.15  21.70  26843405.68   0.21
VIDEO_PLAYERS          4.08   630743.93  15.31  35554301.26   0.06
WEATHER                4.25   178106.52  13.09   5196347.80   0.40

In [37]:
apps.groupby(by='Type')['App'].count()

Type
Free    9591
Paid     765
Name: App, dtype: int64

In [38]:
### Last three columns appear not to be that important. Let's remove them.
apps.drop(columns=['Last Updated', 'Current Ver', 'Android Ver'], inplace=True)

In [40]:
apps.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews  Size  Installs  Type  Price Content Rating  \
0      159  19.0     10000  Free    0.0       Everyone   
1      967  14.0    500000  Free    0.0       Everyone   

                      Genres  
0               Art & Design  
1  Art & Design;Pretend Play